### ISSUE TITLE
`clippy::non_canonical_partial_ord_impl` warns for canonical forwarding implementation

### ISSUE THREAD

**Issue by hrxi** [rust-lang/rust-clippy#14149]
### Summary

.

### Lint Name

clippy::non_canonical_partial_ord_impl

### Reproducer

```rust
use std::cmp::Ordering;

struct A {
    inner: i32,
    cache: i32,
}

impl PartialEq for A {
    fn eq(&self, other: &A) -> bool {
        self.inner == other.inner
    }
}

impl Eq for A {}

impl PartialOrd for A {
    fn partial_cmp(&self, other: &A) -> Option<Ordering> {
        self.inner.partial_cmp(&other.inner)
    }
}

impl Ord for A {
    fn cmp(&self, other: &A) -> Ordering {
        self.inner.cmp(&other.inner)
    }
}
```

Results in this warning:

```
warning: non-canonical implementation of `partial_cmp` on an `Ord` type
  --> 16:1
   |
16 | /  impl PartialOrd for A {
17 | |      fn partial_cmp(&self, other: &A) -> Option<Ordering> {
   | | __________________________________________________________-
18 | ||         self.inner.partial_cmp(&other.inner)
19 | ||     }
   | ||_____- help: change this to: `{ Some(self.cmp(other)) }`
20 | |  }
   | |__^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#non_canonical_partial_ord_impl
   = note: `#[warn(clippy::non_canonical_partial_ord_impl)]` on by default
```

But I expected no warning for this rather canonical forwarding of traits to one particular member.

### Version

```text
rustc 1.84.0 (9fc6b4312 2025-01-07)
binary: rustc
commit-hash: 9fc6b43126469e3858e2fe86cafb4f0fd5068869
commit-date: 2025-01-07
host: x86_64-unknown-linux-gnu
release: 1.84.0
LLVM version: 19.1.5
```

**Comment by ryanavella** [comment]
Note that it misfires even if you change `self.inner.partial_cmp(&other.inner)` to `Some(self.inner.partial_cmp(&other.inner).unwrap())`.

I encountered this misfire myself when forwarding it as `<DerefTarget as PartialOrd>::partial_cmp(self, other)`.

**Comment by Jarcho** [comment]
The implementation should be `Some(self.cmp(&other))` in order to avoid any chance a future change only touches one of them.

**Comment by ryanavella** [comment]
I'm not sure if that was a response to me, but my `Ord` implementation is indeed managed in exactly one place.

```
<T as PartialOrd>                <T as Ord>
         |                            |
         V                            V
<DerefTarget as PartialOrd> ---> <DerefTarget as Ord>
```

**Comment by hrxi** [comment]
> The implementation should be `Some(self.cmp(&other))` in order to avoid any chance a future change only touches one of them.

For forwarded implementations (in smart pointers), it's much more confusing if the `PartialCmp` does not forward to the `PartialCmp` of the pointee.
